README.md
# Quicktime/MP4 Fast Start
Forked from [qtfaststart](https://github.com/danielgtaylor/qtfaststart) which 
has had no maintainer for many years.


Enable streaming and pseudo-streaming of Quicktime and MP4 files by
moving metadata and offset information to the front of the file.

This program is based on qt-faststart.c from the ffmpeg project, which is
released into the public domain, as well as ISO 14496-12:2005 (the official
spec for MP4), which can be obtained from the ISO or found online.

The goals of this project are to run anywhere without compilation (in
particular, many Windows and Mac OS X users have trouble getting
qt-faststart.c compiled), to run about as fast as the C version, to be more
user-friendly, and to use less actual lines of code doing so.

## Features
* Works everywhere Python (2.6+) can be installed
* Handles both 32-bit (stco) and 64-bit (co64) atoms
* Handles any file where the mdat atom is before the moov atom
* Preserves the order of other atoms
* Can replace the original file (if given no output file)
  
## Installing from PyPi

To install from PyPi, you may use ``pip``:

    pip install qtfaststart2

## Installing from source

Download a copy of the source, ``cd`` into the top-level
``qtfaststart2`` directory, and run:

    python setup.py install

If you are installing to your system Python (instead of a virtualenv), you
may need root access (via ``sudo`` or ``su``).

## Usage
See ``qtfaststart2 --help`` for more info! If outfile is not present then
the infile is overwritten::

    $ qtfaststart2 infile [outfile]

To run without installing you can use::

    $ bin/qtfaststart2 infile [outfile]

To see a list of top-level atoms and their order in the file::

    $ bin/qtfaststart2 --list infile

If on Windows, the qtfaststart2 script will not execute, so use::

    > python -m qtfaststart2 ...



CHANGES.md
# Release notes
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 2.1.1
- Readme updates, skipped a few versions here because of an accidental mess-up with PyPi publishing

## 2.0.3
- Fix import errors by renaming everything from qtfaststart to qtfaststart2

## 2.0.2
- Fix: Exit gracefully on already optimised files. See https://github.com/danielgtaylor/qtfaststart/pull/17

## 2.0.1
- Support files with multiple mdat atoms. qtfaststart2 can mistakenly handle files with multiple mdat atoms as already
  set up when the moov atom is actually at the end of the file. This patch makes certain the moov atom is placed before
  all mdat atoms. See https://github.com/danielgtaylor/qtfaststart/pull/19

## 2.0.0
- Create qtfaststart2 which is branched off from [danielgtaylor/qtfaststart](https://github.com/danielgtaylor/qtfaststart)


qtfaststart2/command.py

from __future__ import print_function

import logging
import os
import shutil
import sys
import tempfile

from optparse import OptionParser
from qtfaststart2 import VERSION
from qtfaststart2 import processor
from qtfaststart2.exceptions import FastStartException

log = logging.getLogger("qtfaststart2")

def run():
    logging.basicConfig(level = logging.INFO, stream = sys.stdout,
                        format = "%(message)s")

    parser = OptionParser(usage="%prog [options] infile [outfile]",
                          version="%prog " + VERSION)

    parser.add_option("-d", "--debug", dest="debug", default=False,
                      action="store_true",
                      help="Enable debug output")
    parser.add_option("-l", "--list", dest="list", default=False,
                      action="store_true",
                      help="List top level atoms")
    parser.add_option("-e", "--to_end", dest="to_end", default=False,
                      action="store_true",
                      help="Move moov atom to the end of file")
    parser.add_option("-s", "--sample", dest="sample", default=False,
                      action="store_true",
                      help="Create a small sample of the input file")

    options, args = parser.parse_args()

    if len(args) < 1:
        parser.print_help()
        raise SystemExit(1)

    if options.debug:
        logging.getLogger().setLevel(logging.DEBUG)

    if options.list:
        index = processor.get_index(open(args[0], "rb"))

        for atom, pos, size in index:
            if atom == "\x00\x00\x00\x00":
                # Strange zero atom... display with dashes rather than
                # an empty string
                atom = "----"

            print(atom, "(" + str(size) + " bytes)")

        raise SystemExit

    if len(args) == 1:
        # Replace the original file!
        if options.sample:
            print("Please pass an output filename when used with --sample!")
            raise SystemExit(1)

        tmp, outfile = tempfile.mkstemp()
        os.close(tmp)
    else:
        outfile = args[1]

    limit = 0
    if options.sample:
        # Create a small sample (4 MiB)
        limit = 4 * (1024 ** 2)

    try:
        processor.process(args[0], outfile, limit = limit, to_end = options.to_end)
    except FastStartException:
        # A log message was printed, so exit with an error code
        raise SystemExit(1)

    if len(args) == 1:
        # Move temp file to replace original
        shutil.move(outfile, args[0])


qtfaststart2/processor.py
"""
    The guts that actually do the work. This is available here for the
    'qtfaststart2' script and for your application's direct use.
"""

import shutil
import logging
import os
import struct
import collections

import io

from qtfaststart2.exceptions import FastStartSetupError
from qtfaststart2.exceptions import MalformedFileError
from qtfaststart2.exceptions import UnsupportedFormatError

# This exception isn't directly used, included it for backward compatability
# in the event someone had used it from our namespace previously
from qtfaststart2.exceptions import FastStartException

CHUNK_SIZE = 8192

log = logging.getLogger("qtfaststart2")

# Older versions of Python require this to be defined
if not hasattr(os, 'SEEK_CUR'):
    os.SEEK_CUR = 1

Atom = collections.namedtuple('Atom', 'name position size')

def read_atom(datastream):
    """
        Read an atom and return a tuple of (size, type) where size is the size
        in bytes (including the 8 bytes already read) and type is a "fourcc"
        like "ftyp" or "moov".
    """
    size, type = struct.unpack(">L4s", datastream.read(8))
    type = type.decode('ascii')
    return size, type


def _read_atom_ex(datastream):
    """
    Read an Atom from datastream
    """
    pos = datastream.tell()
    atom_size, atom_type = read_atom(datastream)
    if atom_size == 1:
        atom_size, = struct.unpack(">Q", datastream.read(8))
    return Atom(atom_type, pos, atom_size)


def get_index(datastream):
    """
        Return an index of top level atoms, their absolute byte-position in the
        file and their size in a list:

        index = [
            ("ftyp", 0, 24),
            ("moov", 25, 2658),
            ("free", 2683, 8),
            ...
        ]

        The tuple elements will be in the order that they appear in the file.
    """
    log.debug("Getting index of top level atoms...")

    index = list(_read_atoms(datastream))
    _ensure_valid_index(index)

    return index


def _read_atoms(datastream):
    """
    Read atoms until an error occurs
    """
    while datastream:
        try:
            atom = _read_atom_ex(datastream)
            log.debug("%s: %s" % (atom.name, atom.size))
        except:
            break

        yield atom

        if atom.size == 0:
            if atom.name == "mdat":
                # Some files may end in mdat with no size set, which generally
                # means to seek to the end of the file. We can just stop indexing
                # as no more entries will be found!
                break
            else:
                # Weird, but just continue to try to find more atoms
                continue

        datastream.seek(atom.position + atom.size)


def _ensure_valid_index(index):
    """
    Ensure the minimum viable atoms are present in the index.

    Raise MalformedFileError if not.
    """
    top_level_atoms = set([item.name for item in index])
    for key in ["moov", "mdat"]:
        if key not in top_level_atoms:
            msg = "%s atom not found, is this a valid MOV/MP4 file?" % key
            log.warn(msg)
            raise MalformedFileError(msg)


def find_atoms(size, datastream):
    """
    Compatibilty interface for _find_atoms_ex
    """
    fake_parent = Atom('fake', datastream.tell()-8, size+8)
    for atom in _find_atoms_ex(fake_parent, datastream):
        yield atom.name


def _find_atoms_ex(parent_atom, datastream):
    """
        Yield either "stco" or "co64" Atoms from datastream.
        datastream will be 8 bytes into the stco or co64 atom when the value
        is yielded.

        It is assumed that datastream will be at the end of the atom after
        the value has been yielded and processed.

        parent_atom is the parent atom, a 'moov' or other ancestor of CO
        atoms in the datastream.
    """
    stop = parent_atom.position + parent_atom.size

    while datastream.tell() < stop:
        try:
            atom = _read_atom_ex(datastream)
        except:
            msg = "Error reading next atom!"
            log.exception(msg)
            raise MalformedFileError(msg)

        if atom.name in ["trak", "mdia", "minf", "stbl"]:
            # Known ancestor atom of stco or co64, search within it!
            for res in _find_atoms_ex(atom, datastream):
                yield res
        elif atom.name in ["stco", "co64"]:
            yield atom
        else:
            # Ignore this atom, seek to the end of it.
            datastream.seek(atom.position + atom.size)

def _moov_is_compressed(datastream, moov_atom):
    """
        scan the atoms under the moov atom and detect whether or not the
        atom data is compressed
    """
    # seek to the beginning of the moov atom contents
    datastream.seek(moov_atom.position+8)

    # step through the moov atom childeren to see if a cmov atom is among them
    stop = moov_atom.position + moov_atom.size
    while datastream.tell() < stop:
        child_atom = _read_atom_ex(datastream)
        datastream.seek(datastream.tell()+child_atom.size - 8)

        # cmov means compressed moov header!
        if child_atom.name == 'cmov':
            return True

    return False

def process(infilename, outfilename, limit=float('inf'), to_end=False,
        cleanup=True):
    """
        Convert a Quicktime/MP4 file for streaming by moving the metadata to
        the front of the file. This method writes a new file.

        If limit is set to something other than zero it will be used as the
        number of bytes to write of the atoms following the moov atom. This
        is very useful to create a small sample of a file with full headers,
        which can then be used in bug reports and such.

        If cleanup is set to False, free atoms and zero atoms will not be
        scrubbed from from the mov
    """
    datastream = open(infilename, "rb")

    # Get the top level atom index
    index = get_index(datastream)

    mdat_pos_first = -1
    mdat_pos_last  = -1
    free_size = 0

    # Get position of first and last mdat atom(s)
    for atom in index:
        if atom.name == "mdat":
            if (atom.position < mdat_pos_first or mdat_pos_first < 0):
                mdat_pos_first = atom.position
            if (atom.position > mdat_pos_last or mdat_pos_last < 0):
                mdat_pos_last = atom.position

    if mdat_pos_first < 0:
        # No mdat atom found
        msg = "No mdat atom found."
        log.error(msg)
        raise FastStartSetupError(msg)

    # Make sure moov occurs AFTER mdat, otherwise no need to run!
    for atom in index:
        # The atoms are guaranteed to exist from get_index above!
        if atom.name == "moov":
            moov_atom = atom
            moov_pos = atom.position
        elif atom.name == "free" and atom.position < mdat_pos_first and cleanup:
            # This free atom is before the mdat!
            free_size += atom.size
            log.info("Removing free atom at %d (%d bytes)" %
                    (atom.position, atom.size))
        elif (atom.name == "\x00\x00\x00\x00" and atom.position < mdat_pos_first):
            # This is some strange zero atom with incorrect size
            free_size += 8
            log.info("Removing strange zero atom at %s (8 bytes)" %
                    atom.position)

    # Offset to shift positions
    offset = - free_size
    is_setup = False
    if (moov_pos < mdat_pos_last and to_end):
        # moov is in the wrong place, shift by moov size
        offset -= moov_atom.size
    elif (moov_pos > mdat_pos_first and not to_end):
        # moov is in the wrong place, shift by moov size
        offset += moov_atom.size
    elif free_size == 0:
        # No free atoms to process and moov is correct, we are done!
        msg = "This file appears to be already optimised for streaming. Nothing to do."
        log.info(msg)
        return

    # Check for compressed moov atom
    is_compressed = _moov_is_compressed(datastream, moov_atom)
    if is_compressed:
        msg = "Movies with compressed headers are not supported"
        log.error(msg)
        raise UnsupportedFormatError(msg)

    # read and fix moov
    moov = _patch_moov(datastream, moov_atom, offset)

    log.info("Writing output...")
    outfile = open(outfilename, "wb")

    # Write ftype
    for atom in index:
        if atom.name == "ftyp":
            log.debug("Writing ftyp... (%d bytes)" % atom.size)
            datastream.seek(atom.position)
            outfile.write(datastream.read(atom.size))

    if not to_end:
        _write_moov(moov, outfile)

    # Write the rest
    skip_atom_types = ["ftyp", "moov"]
    if cleanup:
        skip_atom_types += ["free"]

    atoms = [item for item in index if item.name not in skip_atom_types]
    for atom in atoms:
        log.debug("Writing %s... (%d bytes)" % (atom.name, atom.size))
        datastream.seek(atom.position)

        # for compatability, allow '0' to mean no limit
        cur_limit = limit or float('inf')
        cur_limit = min(cur_limit, atom.size)

        for chunk in get_chunks(datastream, CHUNK_SIZE, cur_limit):
            outfile.write(chunk)

    if to_end:
        _write_moov(moov, outfile)

    # Close and set permissions
    outfile.close()
    try:
        shutil.copymode(infilename, outfilename)
    except:
        log.warn("Could not copy file permissions!")

def _write_moov(moov, outfile):
    # Write moov
    bytes = moov.getvalue()
    log.debug("Writing moov... (%d bytes)" % len(bytes))
    outfile.write(bytes)

def _patch_moov(datastream, atom, offset):
    datastream.seek(atom.position)
    moov = io.BytesIO(datastream.read(atom.size))

    # reload the atom from the fixed stream
    atom = _read_atom_ex(moov)

    for atom in _find_atoms_ex(atom, moov):
        # Read either 32-bit or 64-bit offsets
        ctype, csize = dict(
            stco=('L', 4),
            co64=('Q', 8),
        )[atom.name]

        # Get number of entries
        version, entry_count = struct.unpack(">2L", moov.read(8))

        log.info("Patching %s with %d entries" % (atom.name, entry_count))

        entries_pos = moov.tell()

        struct_fmt = ">%(entry_count)s%(ctype)s" % vars()

        # Read entries
        entries = struct.unpack(struct_fmt, moov.read(csize * entry_count))

        # Patch and write entries
        offset_entries = [entry + offset for entry in entries]
        moov.seek(entries_pos)
        moov.write(struct.pack(struct_fmt, *offset_entries))
    return moov

def get_chunks(stream, chunk_size, limit):
    remaining = limit
    while remaining:
        chunk = stream.read(min(remaining, chunk_size))
        if not chunk:
            return
        remaining -= len(chunk)
        yield chunk
